### ISSUE TITLE
specify the cores to be used by wrapped jobs

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#316]
In GitLab by @ editomas on Mar 16, 2018, 12:01

@ lbatista, @dmanubens in the current implementation of the wrapper the different wrapped jobs are using the same cores to run `mpirun`.

Support has checked this in this test we made `/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu/ASThread_15210377404757_4632_3.cmd`: 

```
/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu> cat a0xu_20120201_009_1_SIM.cmd.0.out a0xu_20120201_010_1_SIM.cmd.1.out
a0xu_20120201_011_1_SIM.cmd.2.out | sort -n | uniq |grep -e "^s[0-9][0-9]"
```

As you know, they have explained that this happens because the `#SBATCH` are only processed with the `sbatch` command, and hence for the wrapper job only. The wrapped jobs are using `mpirun` without processing their individual `#SBATCH` instructions, and so take all the same first cores.

As they have suggested, a solution to this would to call the different mpirun with machinefile flags:

```
mpirun -machinefile list1 ./job1 
mpirun -machinefile list2 ./job2 
...
```

Maybe our old way to submit ensemble simulations in parallel can be of help. Please see:
https://earth.bsc.es/gitlab/ac/nmmb-monarch/blob/develop-DA-parallel/JOB/DUST_DA/submit_ENS_parallel.cmd

It was using `$SLURM_JOB_NODELIST` (or `$LSB_MCPU_HOSTS` for `LSF`) to access the list of cores.

Additionally, you'll see that it was taking the I/O cores all from the same nodes, in order to optimize the usage of resources.

**Comment by kinow** [comment]
In GitLab by @ lbatista on Mar 16, 2018, 12:05

Hi @ editomas thank you for documenting it, I have just started working on it and will keep you posted!
FYI: @mcastril

**Comment by kinow** [comment]
In GitLab by @ editomas on Mar 16, 2018, 17:16

@ lbatista here you have a cleaner version of the script for `SLURM`, and with an example of the temporary files and machine files:

`/gpfs/projects/bsc32/bsc32137/machine_files/`

As we said, in the simulation scripts (eg `a0xu_20120201_002_1_SIM.cmd`) `mpirun` will have to be run with:

`mpirun -np ${NUMPROC} -machinefile ${machine_file_path}"machine_file_2" NMMB.`x

Please let me know if you will need it for LSF in the future. I have only added an example for it:

`/gpfs/projects/bsc32/bsc32137/machine_files/example_LSF_machine_file_0`

**Comment by kinow** [comment]
In GitLab by @ lbatista on Mar 16, 2018, 19:15

I changed the `ASThread .. cmd` script to get the list of nodes, create a folder called machinefiles, split the list of nodes into sublists for each member inside the wrapper in the format machinefile_MEMBER.

This part is not generic enough as it relies on the name of the job being `EXPID_DATE_MEMBER_CHUNK`, so this should be improved as soon as possible. However, it should do the trick for the moment.

The mpirun inside the SIM script still needs to be changed, but I assume it's just a matter of changing it to the following?

``` 
mpirun -np ${NUMPROC}  -machinefiles/machinefile_${MEMBER} list-first-1500 ${UMO_OUT}/NMMB.x >& $> {log_file} 
```


Also, I haven't separated the nodes for computational and write tasks. Do you think it's worth a try without doing it?

**Comment by kinow** [comment]
In GitLab by @ editomas on Mar 16, 2018, 19:26

@ lbatista I could make a first test with it. Can you give me a version of this:
`/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu/ASThread_15210377404757_4632_3.cmd` 
with your changes? I'll change the SIM scripts and run it.

**Comment by kinow** [comment]
In GitLab by @ lbatista on Mar 16, 2018, 19:38

Since this is only the horizontal, I think it's easier to use `machinefile_${JOBNAME}` instead of MEMBER that will be needed for the hybrid

```

#!/usr/bin/env python
###############################################################################
#              ASThread_15210377404757_4632_3
###############################################################################
#
#SBATCH -J ASThread_15210377404757_4632_3
#
#SBATCH -A pr1edr00
#SBATCH -o ASThread_15210377404757_4632_3.out
#SBATCH -e ASThread_15210377404757_4632_3.err
#SBATCH -t 00:05:00
#SBATCH -n 4632
#

#
###############################################################################

import os
import sys
from threading import Thread
from commands import getstatusoutput

class JobThread(Thread):
    def __init__ (self, template, id_run):
        Thread.__init__(self)
        self.template = template
        self.id_run = id_run

    def run(self):
        out = str(self.template) + "." + str(self.id_run) + ".out"
        err = str(self.template) + "." + str(self.id_run) + ".err"
        command = "bash " + str(self.template) + " " + str(self.id_run) + " " + os.getcwd()
        (self.status) = getstatusoutput(command + " > " + out + " 2> " + err)

# Getting the list of allocated nodes
os.system("scontrol show hostnames $SLURM_JOB_NODELIST > node_list")
os.system("mkdir machinefiles")
# Splits the list of nodes into sublists with the number of nodes necessary for each job and 
# adds it to the folder machinefiles in the format machinefile_XX
os.system("cat node_list | split -a 2 -d -l 33  - machinefiles/machinefile_")

# Defining scripts to be run
scripts = ['a0xu_20120201_009_1_SIM.cmd', 'a0xu_20120201_010_1_SIM.cmd', 'a0xu_20120201_011_1_SIM.cmd']

# Initializing PIDs container
pid_list = []

# Initializing the scripts
for i in range(len(scripts)):
    machinefile = 'machinefile_' + format(id, '02d')
    job = scripts[i]
    rename_cmd = "mv machinefiles/"+machinefile+" machinefiles/machinefile_"+job.replace(".cmd", '')
    os.system(rename_cmd)
    current = JobThread(job, i)
    pid_list.append(current)
    current.start()

# Waiting until all scripts finish
for pid in pid_list:
    pid.join()
    completed_filename = scripts[i].replace('.cmd', '_COMPLETED')
    completed_path = os.path.join(os.getcwd(), completed_filename)
    if os.path.exists(completed_path):
        print "The job ", pid.template," has been COMPLETED"
    else:
        print "The job ", pid.template," has FAILED"
```

**Comment by kinow** [comment]
In GitLab by @ editomas on Mar 19, 2018, 01:44

@ lbatista thanks for this, I've tested it with `/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu/ASThread_15210377404757_4632_3_MF.cmd`. I wonder if the format of the machine files should be different and the name of the nodes should be repeated for each core. If you see now `/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu/machinefiles/machinefile_a0xu_20120201_009_1_SIM` has a list of 33 nodes, while `/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu/machinefiles/machinefile_a0xu_20120201_011_1_SIM` has a list of 31 nodes. In fact I could complete only the first two of the three simualitons. The way we were doing it in the past was to have (in this example) 1544 rows in each machinefile for the respective cores to use, with each row containing the names of the nodes to which the cores belong. E.g.,

`s02r1b41
s02r1b41
s02r1b41
...
s18r3b08`

This way the same node can be shared among simulations. This was working without problems in MN4 (we did not need to have multiple of 48 cores for a simulation). I wrote to support to know if they think that having different mpirun accessing the same node now creates any conflict. Let's see what they say.

A small thing: in your script `i` instead of `id` should go in `format(id, '02d')`

**Comment by kinow** [comment]
In GitLab by @ lbatista on Mar 19, 2018, 13:03

Hi @ editomas 

Would it be a problem changing the number of cores to 1536 or 1584? It seems the simplest way to go.

If it's a problem, we can try specifying the cores used for each node. Could you please try this:

```
#!/usr/bin/env python
###############################################################################
#              ASThread_15210377404757_4632_3
###############################################################################
#
#SBATCH -J ASThread_15210377404757_4632_3
#
#SBATCH -A pr1edr00
#SBATCH -o ASThread_15210377404757_4632_3.out
#SBATCH -e ASThread_15210377404757_4632_3.err
#SBATCH -t 00:05:00
#SBATCH -n 4632
#

#
###############################################################################

import os
import sys
from threading import Thread
from commands import getstatusoutput

class JobThread(Thread):
    def __init__ (self, template, id_run):
        Thread.__init__(self)
        self.template = template
        self.id_run = id_run

    def run(self):
        out = str(self.template) + "." + str(self.id_run) + ".out"
        err = str(self.template) + "." + str(self.id_run) + ".err"
        command = "bash " + str(self.template) + " " + str(self.id_run) + " " + os.getcwd()
        (self.status) = getstatusoutput(command + " > " + out + " 2> " + err)

# Getting the list of allocated nodes
os.system("scontrol show hostnames $SLURM_JOB_NODELIST > node_list")
os.system("mkdir machinefiles")

# Defining scripts to be run
scripts = ['a0xu_20120201_009_1_SIM.cmd', 'a0xu_20120201_010_1_SIM.cmd', 'a0xu_20120201_011_1_SIM.cmd']

# Initializing PIDs container
pid_list = []

with open('node_list', 'r') as file:
     all_nodes = file.read()
 
all_nodes = all_nodes.split('\n')
remaining_cores = 0

# Initializing the scripts
for i in range(len(scripts)):
	job = scripts[i]
	total_cores = ({4} / len(scripts))
	machines = str()                
                
	for i in reversed(range(len(all_nodes))):
        	node = all_nodes[i]
        	if total_cores >= 48:
                	if remaining_cores > 0:
                        	machines += node + ':' + str(remaining_cores) +'\n'
                        	total_cores -= remaining_cores
                        	remaining_cores = 0
                	else:
                            machines += node + '\n'
                            total_cores -= 48
                            all_nodes.pop(i)
        	else:
                	remaining_cores = 48 - total_cores
                	machines += node + ':' + str(total_cores) +'\n'
                	break
    machines = "\n".join([s for s in machines.split("\n") if s])
    with open("machinefiles/machinefile_"+job.replace(".cmd", ''), "w") as machinefile:
        machinefile.write(machines)
    current = JobThread(job, i)
    pid_list.append(current)
    current.start()

# Waiting until all scripts finish
for pid in pid_list:
    pid.join()
    completed_filename = scripts[i].replace('.cmd', '_COMPLETED')
    completed_path = os.path.join(os.getcwd(), completed_filename)
    if os.path.exists(completed_path):
        print "The job ", pid.template," has been COMPLETED"
    else:
        print "The job ", pid.template," has FAILED"
```

About i instead of id, you are right. id is for the case of the hybrid wrapper, so I forgot to change when "translating" to the horizontal, sorry for the typo.

**Comment by kinow** [comment]
In GitLab by @ editomas on Mar 19, 2018, 16:47

@ lbatista, In this case we cannot change the number of cores per simulation to 1536 or 1584 becasue, if I'm not mistaken (I'm writing in a hurry, sorry), we cannot find two numbers such that they are equal to (INPES*JNPES)+WRTSK, with WRTSK usually =8 (or =4 for coarser resolutions). A possible close alternative would be 2976 = (37*40)+8

So, it is not a problem, but we would loose some flexibility. @orioljorba , do you see it a problem to have to use always a number of cores multiple of 48 for MONARCH simulations? 

I would suggest that we try to find a general solution that let us choose also non multiple of 48. I'm testing your new version, let's see if it works.

Also, this solution of running mpirun with machine files that you are implementing would not be useful also for other applications to be wrapped together that might need any number of cores? what about applications that would need less then 48 cores, for example?

thanks for helping!

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 19, 2018, 16:54

Hi @ lbatista , maybe is possible to ask for a MAX_CORES_PER_NODE variable in your script? We can talk about it.

**Comment by kinow** [comment]
In GitLab by @ lbatista on Mar 19, 2018, 17:24

Hi, 

if this solution works, it should be generic for wraps/jobs of any cores. @ editomas I am afraid when writing the code directly here, the identation in some parts gets a bit messed up. I checked the .cmd in your experiment and corrected the identation. Please find attached the file [ASThread_15210377404757_4632_3_MF2.cmd](https://earth.bsc.es/gitlab/es/autosubmit/uploads/f21a3ee0037ad859b0f8f06dbae227b9/ASThread_15210377404757_4632_3_MF2.cmd). Let me know how it goes

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 19, 2018, 17:42

According to Intel documentation you indeed need one line per core, or otherwise group some lines with the semicolon (:):

(chapter Machine File Specification)

https://software.intel.com/en-us/articles/controlling-process-placement-with-the-intel-mpi-library

In any case you can always see in the .out file how the result is, by using `export I_MPI_DEBUG=5` before calling mpirun.

**Comment by kinow** [comment]
In GitLab by @ lbatista on Mar 19, 2018, 17:48

Thanks @mcastril, @ editomas can you please add it to the SIM script when testing?

**Comment by kinow** [comment]
In GitLab by @ editomas on Mar 20, 2018, 14:44

The fact that simulations are sharing nodes gives still error when loading the module fabric (as far as understand this module is a fix by support to be able to run our model in MN4). In /`gpfs/scratch/pr1edr00/pr1edr02/a0xu/nmmb-monarch/OUTPUT/REG/003/CURRENT_RUN/nmm_rrtm.out_1` in fact there is the error:

```
s22r2b61.139634PSM2 can't open hfi unit: -1 (err=23)
[258] ofi netmod(): ofi_opendomain error: ret -5, Input/output error
[258] MPI startup(): ofi fabric is not available and fallback fabric is not enabled
```

and in fact the s22r2b61 node is shared with another simulation.

I had already written to support my concern that "two mpirun can end up using the same node but different cores". I will write to them again, but I already suspect that we might have to do as @ lbatista said and use only multiple of 48. Larissa, sorry if we did not go directly for that solution. 

a smaller issue: 
`/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu/ASThread_15210377404757_4632_3_MF2_2smalljobs.out`
says that the two simulations were completed even if there is a complete file only for one of them

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 20, 2018, 16:01

Hi @ editomas , I don't understand very well why loading the module fabric and sharing nodes could lead to an error...

I suggested @ lbatista to look if "set -u" or "set -e" in the script could be guilty of this behavior. Otherwise I don't see any problem but if this is an issue operations should fix it.

Have you been able to check in the .out if processes are being correctly assigned?

**Comment by kinow** [comment]
In GitLab by @ editomas on Mar 20, 2018, 16:47

@mcastril I'll forward you the emails that I have exchanged with support about this error with the "module load fabric" and wrapped simulations. They say something like this: "The error is produced because somebody is using already the hfi ports, and in this case if 2 mpiruns are trying to use the same nodes". We could test to run "module load fabric" in the wrapper job (and hence only once), instead of having this in each of the wrapped simulation scripts (we just have to be patient with the long queue times today). If you think that it is a useless test, just let me know it.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 20, 2018, 17:06

Thanks @ editomas ,

What I have read in the emails is that all your small jobs where going to the same node, so not only the jobs where sharing a node, but also they were trying to get the same cores (because every job was trying to use 48 cores of the same node).

But that does not explain while you should have any problem if sharing nodes but not sharing cores. For example, if in the same node you use 24 cores for one small job and 24 for another one.

The first is a problem of oversubscribing, and the second would be more strange, that's why I was asking if you are sure of where the different small jobs are being run. Using I_MPI_DEBUG=5 you can get a nice output like this:

```
[0] MPI startup(): Rank    Pid      Node name  Pin cpu
[0] MPI startup(): 0       321976   s03r1b43   {0,1,2,3,4,5}
[0] MPI startup(): 1       321977   s03r1b43   {6,7,8,9,10,11}
[0] MPI startup(): 2       321978   s03r1b43   {12,13,14,15,16,17}
[0] MPI startup(): 3       321979   s03r1b43   {18,19,20,21,22,23}
[0] MPI startup(): 4       321980   s03r1b43   {24,25,26,27,28,29}
[0] MPI startup(): 5       321981   s03r1b43   {30,31,32,33,34,35}
[0] MPI startup(): 6       321982   s03r1b43   {36,37,38,39,40,41}
[0] MPI startup(): 7       321983   s03r1b43   {42,43,44,45,46,47}
```

Where you can know the location of all your ranks within the nodes and CPUs.

Aside from this, I see that when David was speaking about the problem with fabric module, was before he knew your jobs where taking the same nodes (and cores). So maybe the module has nothing to be with this issue.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 20, 2018, 17:17

Ok we saw that you included the DEBUG variable :) . We are trying to get the output but we need read access to /`gpfs/scratch/pr1edr00/pr1edr02/a0xu/nmmb-monarch/OUTPUT/REG/` and subfolders.

**Comment by kinow** [comment]
In GitLab by @ editomas on Mar 20, 2018, 17:36

changed the permission with chmod -R a+r and just in case also a+w, in case you still cannot see them in the prace account, a copy is here:

failed simulation:
`/gpfs/projects/bsc32/bsc32137/nmm_rrtm.out_1_secondSIMfails`

not failed simulation:
`/gpfs/projects/bsc32/bsc32137/nmm_rrtm.out_1`

the list of MPI startup() is only in the second one. the first stops earlier.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 20, 2018, 17:56

Thanks @ editomas , I see, as you say, that it crashes before the mpirun is able to output nothing... I will take a look at the other log files.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 20, 2018, 18:06

Hi @ editomas , you should give -R a+X so we cannot lit the directory contents :)

**Comment by kinow** [comment]
In GitLab by @ editomas on Mar 20, 2018, 18:53

ok applied also execution permission, sorry. please note that in this last test I run only two simulations (002 and 003, to wait less time in the queue).
the log for the not failed simulation gets moved from the CURRENT_RUN dir and goes in: ´/gpfs/scratch/pr1edr00/pr1edr02/a0xu/nmmb-monarch/OUTPUT/REG/002/2012020100/nmm_rrtm.out_1´

while for the failed simulation stays in:
´/gpfs/scratch/pr1edr00/pr1edr02/a0xu/nmmb-monarch/OUTPUT/REG/003/CURRENT_RUN/nmm_rrtm.out_1´

I can see in the log that the first simulation is stealing cores that should be used by the second (what a thief!): is using 48 cores for  s22r2b61 while the machine file ´/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu/machinefiles/machinefile_a0xu_20120201_002_1_SIM_secondSIMfails´ says: 
s22r2b61:24
thanks @mcastril for the DEBUG flag.

there is in the queue since a while a new test with the name of the node repeated for each task (thanks @ lbatista). I'll let you know how it goes with that format of machine files

**Comment by kinow** [comment]
In GitLab by @ editomas on Mar 21, 2018, 11:08

also the new format of machine files gives the same error for the shared node: 
`s05r1b14.73116PSM2 can't open hfi unit: -1 (err=23)` 

It crashes before mpirun is able to output the list of cores: 
```
/gpfs/scratch/pr1edr00/pr1edr02/a0xu/nmmb-monarch/OUTPUT/REG/002/CURRENT_RUN/nmm_rrtm.out_1_bothSIMfail
/gpfs/scratch/pr1edr00/pr1edr02/a0xu/nmmb-monarch/OUTPUT/REG/003/CURRENT_RUN/nmm_rrtm.out_1_bothSIMfail
```

s05r1b14 in theory should be shared between the two simulations (24 cores each):
```
/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu/machinefiles/machinefile_a0xu_20120201_002_1_SIM_bothSIMfail
/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu/machinefiles/machinefile_a0xu_20120201_003_1_SIM_bothSIMfail
/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu/node_list_bothSIMfail
```

I'll show this to support with you in copy. The test with the loading of modules in the wrapper job is in the queue.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 21, 2018, 12:16

Hi,

Could it be that the system is so dumb that each mpirun is using 24 cores from the same node, but the same 24 cores and leaving 24 free? 

Let's see what Operations say about that. Thanks for the reports @ editomas

**Comment by kinow** [comment]
In GitLab by @ editomas on Mar 21, 2018, 12:36

@mcastril with loading the modules in the wrapper, somehow one simulation went through (though the log shows this message  `MPID_nem_ofi_init(): cannot load default ofi library libfabric.so`), and we can read which cores it gets. Can you please have a look at these files:

here it says that s03r2b31 should use 24 cores:
`/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu/machinefiles/machinefile_a0xu_20120201_005_1_SIM_load_wrapper`

here it says that instead it is taking 48 cores from s03r2b31:
`/gpfs/scratch/pr1edr00/pr1edr02/a0xu//nmmb-monarch/OUTPUT/REG/005/2012020100/nmm_rrtm.out_1`

so, it seems that mpirun is not doing what the machine files tell.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 21, 2018, 12:56

Hi @ editomas , the message you are putting there is kind of "normal" (I see it in all the simulation). I will look at the files.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 21, 2018, 13:04

Yes, it seems that it does not pay attention to the machinefiles.

In any case, one thing I see is that the machinefile present the list of nodes in reverse order. Could it be that it is getting confused because of that?

I am also surprised to see that in the .out , it is only using 24 cores from the first node in the list (s03r2b25), while normally I think it's the last node, the one which remains with free cores if you run less cores than requested.  

So the normal procedure is start filling from the first node to the last node, which maybe can have less allocated cores than the cores it has available.

So maybe it's not that is not reading the machinefile, but it's interpreting it wrongly (only a theory).

@ lbatista would it be possible to have the machinefile ordered in alphabetical way?

**Comment by kinow** [comment]
In GitLab by @ lbatista on Mar 21, 2018, 13:07

yes, it's possible, I will order the nodes alphabetically

**Comment by kinow** [comment]
In GitLab by @ editomas on Mar 21, 2018, 16:37

It works! (job `/gpfs/scratch/pr1edr00/pr1edr02/a0xu/LOG_a0xu/ASThread_15210377404757_4632_3_MF2_2smalljobs_order.out.`)

Now we can give ourselves a pat on the shoulder. @ lbatista let me know when I can test this with the ens-wrapper-test branch. I got the simulations to run with that branch without wrapper, but I get an error in the POST that I have to look at.

I close this issue, just to switch to issue #312 ;)

**Comment by kinow** [comment]
In GitLab by @ editomas on Mar 21, 2018, 16:38

closed

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 3, 2018, 17:01

reopened

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 3, 2018, 17:31

Following what I was writing to some of you you in Trello, this was our original script to create machinefiles with LSF:   `https://earth.bsc.es/gitlab/ac/nmmb-monarch/blob/develop-DA-parallel/JOB/DUST_DA/submit_ENS_parallel.cmd`

To solve this issue, I had made here a cleaner version that works for SLURM: 
/`gpfs/scratch/bsc32/bsc32137/machine_files/create_machine_files_for_ENS_parallel.cmd`
there is an example of the machinefiles with I/O and computing cores treated differently. 

I understand that we first have to be sure of the reason why for MONARCH "`mpirun -np 204`" takes always 16 minutes (unless MN4 under-performs) and "`time mpirun -np 204 -machinefile list_of_204_cores`" takes 40-45 minutes (and the figure varies from run to run). 

As @mcastril said, it could not be related to I/O cores.

Do you think I should contact support first? 

Or can we learn something if I run a test with 
`mpirun -np 204 hostname` before the mpirun?

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 3, 2018, 17:32

Hi,

So as @ editomas mentioned, it would be good to compare how the cores are ditributed by using the machinefile or not. This can be done by using the I_MPI_DEBUG environment variable and looking at the .out logfile.

If the distribution is exactly the same, then we would have to see if some other default options are overwritten when used the machinefiles, as the core pinning.

**Comment by kinow** [comment]
In GitLab by @ lbatista on Jul 3, 2018, 17:34

do you think it could be worth testing the first simulations (the ones in 20120101_000_1_SIM ... a156_20120101_012_1_SIM) in a horizontal wrapper to see how long they take to complete?

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 3, 2018, 17:37

Hi @ editomas , to be more precise, I do not discard it could be related to I/O cores. The key is why the behavior is different with or without machinefiles, given that both cases distribute the cores in a sequential order AFAIK.

As I said in my previous message (we wrote almost at the same time) I will run both 
`mpirun -np 204` and `mpirun -np 204 -machinefile list_of_204_cores` and then would see how the cores are distributed in the .out logfile.

Maybe you already have a log from your recent runs that we can take a look into.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 3, 2018, 17:41

In order to see that info remember to declare `export I_MPI_DEBUG=5`

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 3, 2018, 17:48

@mcastril for the second case here is an example of the machinefiles for the different simulations: 
`/gpfs/projects/bsc32/bsc32137/example_machinefiles_thirdtest`
but these are as expected: autosubmit creates them this way. For the first case (mpirun -np 204), I need to run a test where I save the information in the log as you said. I'll do it.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 3, 2018, 17:52

Thanks @ editomas , for both cases anyway we would need the info provided by the runtime. One thing is how the machinefile is created (which is useful for sure) and one another how mpirun is using that machinefile.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 3, 2018, 18:22

Hi @ editomas we saw you are doing some tests but we don't see the debug information, besides you are exporting the I_MPI_DEBUG variable. Maybe the problem is that you are loading the impi module afterwards.

Could you set I_MPI_DEBUG near the mpirun call?

At the same time we don't know where the machinefiles are being stored. Could you give us access to them? :)

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 3, 2018, 18:34

the tests are in prace, I think you don't have access to it (or am I wrong?). I'm copying here files for you: `/gpfs/projects/bsc32/bsc32137/FOR_MC`

I've done the test with `mpirun -np 204`

`/gpfs/projects/bsc32/bsc32137/FOR_MC/nmm_rrtm.out_0` shows the cores used. 

I see that it does:

```
node 1: cores 0-20 24-43
node 2: cores 0-20 24-43
node 3: cores 0-20 24-43
node 4: cores 0-20 24-43
node 5: cores 0-19 24-43
```

so, for some reasons, it leaves out always the same cores (21-23 and 44-47).

the second test is queuing, since it has to be done at running time, I had to run the full wrapper. 

is it worth running test number 1 with a different number of cores and send you the nmm_rrtm.out_0?

please let me know if I can give you more information or permissions to the files.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 3, 2018, 18:43

yes if you see here: /gpfs/projects/bsc32/bsc32137/FOR_MC/a156_20120214_000_2_SIM.cmd
I'm doing the export at the beginning of the script, while modules are loaded afterwards with  . ${EXP_ROOT}/remoteconfig.${ARCH} (line 196). but I think it worked as I see debug info. please let me know if I'm wrong and I'll do it again.

the machine files are created here at each run
/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/machinefiles/

Since I don't think you can access them, I've saved an example in:
/gpfs/projects/bsc32/bsc32137/example_machinefiles_thirdtest

I'll make a copy for you also of test number 2 with the debug option

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 3, 2018, 18:43

Hi @ editomas , I will take a look. I have access to pr1egi00 and your scratch folder (including LOG directory) but I could not see some folders in NMMB structure.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 3, 2018, 18:45

Hi again, I cannot access to /gpfs/projects/bsc32/bsc32137/FOR_MC/nmm_rrtm.out_0 neither.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 3, 2018, 18:47

ooh yeah, can you try now?

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 3, 2018, 18:56

Thanks now I can see the info :)

We were looking at the `/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/a156_20120214_000_2_SIM.20180703150540_test.out` file before and we didn't see any debug info, that's why I was worried about the order of the export/modules.

As you say, it is using only 0-20 and 24-43. That is because there are 24 physical cores per socket and it is using them equally, leaving 4 cores per socket free (as you are requesting 40 cores per node).

Which of the machinefiles would correspond to this test?

I see `a156_20120214_000_2_SIM.cmd` in the same folder as the out file, so I guess `machinefile_a156_20120214_002_1_SIM` would be the hostfile?

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 3, 2018, 18:59

The stamps for the machinefiles are a little bit old so maybe these are not from this test. Maybe the easiest way is that you grant me access to the folder in pr1egi00 so you don't have to copy files.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 3, 2018, 19:02

I will leave now but tomorrow we can continue taking a look.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 3, 2018, 19:31

where am I requesting 40 cores per node? I don't think we are asking that anywhere, but let me know if I'm missing this point.

the files you looked at are for test number 1, i.e. no machinefiles, only `mpirun -np 204` hence there is not machinefile corresponding to this test.

the test number 2 with machinefiles is still queueing, you have to be patient ;).

regarding permissions, I have done:

pr1egi09@login1:/gpfs/scratch/pr1egi00/pr1egi09> chmod -R g+rwX a156

pr1egi09@login1:/gpfs/scratch/pr1egi00/pr1egi09> chmod -R g+rw a156

now I'll do for the full account:

/gpfs/scratch/pr1egi00> chmod -R g+rwX pr1egi09

/gpfs/scratch/pr1egi00> chmod -R g+rw pr1egi09

let me know if this is not enough. happy to grant full permissions to everything.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 4, 2018, 10:36

[This message can wait after dev day]

An example of 3 log files to see how the cores are distributed when using machinefiles are here:

/esarchive/exp/monarch/a156/original_files/000/2012010200/nmm_rrtm.out_0

/esarchive/exp/monarch/a156/original_files/000/2012021400/nmm_rrtm.out_0

/esarchive/exp/monarch/a156/original_files/000/2012021500/nmm_rrtm.out_0

(there are more example if you change i the path 000 with 001, 002, ... 011)

the respective 3 machinefiles are:

/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/machinefiles/machinefile_a156_20120101_000_2_SIM

/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/machinefiles/machinefile_a156_20120214_000_1_SIM

/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/machinefiles/machinefile_a156_20120214_000_2_SIM

(next test I will not re-direct the logs of MONARCH executable, so that you can find them in the expected log of the job)

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 4, 2018, 10:39

Thanks @ editomas 

> where am I requesting 40 cores per node? I don't think we are asking that anywhere, but let me know if I'm missing this point.

Sorry I didn't check that. I guessed that you were doing so because 204 is not multiple of 48. 204 doesn't fit in only 4 nodes, so by asking 5 nodes the distribution is of ~40 cores per node (41:41:41:41:40).

> 
> the files you looked at are for test number 1, i.e. no machinefiles, only `mpirun -np 204` hence there is not machinefile corresponding to this test.

My fault again, I didn't remember you were only submitting the no-machinefile case.

> 
> the test number 2 with machinefiles is still queueing, you have to be patient ;).
> 
> regarding permissions, I have done:
> 
> pr1egi09@login1:/gpfs/scratch/pr1egi00/pr1egi09> chmod -R g+rwX a156
> 
> pr1egi09@login1:/gpfs/scratch/pr1egi00/pr1egi09> chmod -R g+rw a156
> 
> now I'll do for the full account:
> 
> /gpfs/scratch/pr1egi00> chmod -R g+rwX pr1egi09
> 
> /gpfs/scratch/pr1egi00> chmod -R g+rw pr1egi09
> 
> let me know if this is not enough. happy to grant full permissions to everything.

I will check and let you know.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 4, 2018, 10:53

ok. I'll make a test with a multiple of 48 cores and let you know

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 4, 2018, 13:22

the test with multiple of 48 cores (8x23+8=192 for each SIM) shows that the simulation is run efficiently (similar time to when no machinefile are used). An example of log file and machine file:

`/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/a156_20120214_000_1_SIM.cmd.1.out`

`/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/machinefiles/machinefile_a156_20120214_000_1_SIM`

Execution time: 17m45.773s

This is one example, but there are more in the same directories if you look at the most recent files.

Do have to keep this constrain of using multiple of 48 cores or the distribution can be corrected in autosubmit?

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 4, 2018, 13:34

In any case, and this is another story but, it shouldn't have so much impact to move some cores to other nodes. On the contrary, the processes on nodes using only 36 or 24 cores in the previous case should run faster because there is more memory bandwidth available. 

Could you point me to the .err files of the experiments you were mentioning in this message¿?

https://earth.bsc.es/gitlab/es/autosubmit/issues/316#note_23376

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 4, 2018, 13:47

I have them for the last two I had mentioned in the comment:

/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/a156_20120214_000_1_SIM.20180703182458.err

grep real a156_20120214_000_1_SIM.20180703182458.err
real	43m5.591s

/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/a156_20120214_000_2_SIM.cmd.9.err

/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156> grep real a156_20120214_000_2_SIM.cmd.9.err
real	42m56.499s

(the first I had mentioned has been rewritten this morning by the new test, since somehow there was not timestamp in its name).

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 4, 2018, 15:20

Hi @ editomas . Was `/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/machinefiles/machinefile_a156_20120214_000_1_SIM` overwritten?

There I see:

```
uniq -c /gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/machinefiles/machinefile_a156_20120214_000_1_SIM
     48 s05r1b65
     48 s05r1b66
     48 s05r1b67
     48 s06r2b11
```

But in `/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/a156_20120214_000_1_SIM.20180703182458.err`

I see 204 cores

`mpirun -np 204 -machinefile /gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/machinefiles/machinefile_a156_20120214_000_1_SIM /gpfs/scratch/pr1egi00/pr1egi09/a156/nmmb-monarch/OUTPUT/REG/20120214/000/CURRENT_RUN/NMMB.x`

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 4, 2018, 15:50

yes, it has been overwritten by this morning test with 192 cores. the log with 204 cores for which you have asked me earlier the error files was from yesterday test. 

please let me know if you want that I re-run the test with 204 cores and save the machinefiles. I'm happy to do that.

the error file for the test with 192 core is instead:

`/gpfs/scratch/pr1egi00/pr1egi09/a156/LOG_a156/a156_20120214_000_1_SIM.cmd.1.err`

in fact, it says: real	17m45.773s

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 4, 2018, 15:58

Hi @ editomas , by the moment I asked Kim to include the I_MPI_DEBUG flag in the operational run, just to see how the cores are distributed there.

I think the generalized thinking is that leaving the I/O processes alone in one node is better for NMMB performance. But we have seen that the more imbalance is the workload, the worse the model performs. On the other hand when no machinefile is used, mpirun tries to distribute evenly the processes among the nodes. If that's the case, the operational model would be running with a balanced distribution, and not with a separation between compute and I/O nodes, as it was in MN3.

By the way, how many I/O processes are you using? It would be good to know how the imbalance is affecting NMMB performance, and why the execution takes so long when some nodes are half-empty and others not.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 4, 2018, 16:19

In all these tests I've used 8 cores for I/O (WRTSK=8), for lower resolution we can use also 4. just to add for information, in the configuration of MONARCH we specify the cores for the i (INPES) and j (JNPES) direction and for I/O, hence if we have to impose a multiple of 48 we have to be able to decompose total_PROCESSORS-WRTSK=INPES*JNPES 

please let us know what you discover further, and thanks for helping.

**Comment by kinow** [comment]
In GitLab by @ lbatista on Jul 17, 2018, 19:19

@mcastril @ editomas a few things here:
1.  Miguel, did you have a chance to check how the cores are distributed in the operational run?
2.  I was discussing with Enza today to try to understand what options we have. The first thing would be to determine the number of cores to be used for the SIM. The main constraint is having a similar wallclock/duration as the DA. Regarding the machinefiles, basically I see 2 "quick" options: 
    1. The most straightforward would be to have a number of cores multiple of 48, keeping in mind that this number needs to be decomposed into `PROCESSORS-WRTSK=INPES*JNPES`, as mentioned above. 
    2. The alternative is to specify how many cores to use per node (with the constraint of **using the same number of cores per node**, as in the DA) by setting the TASKS variable in the job configuration. In this case, we would request full nodes, and if TASKS < 48, the cores not used in each node would be wasted. For example, if we have PROCESSORS = 205, we would request 240 cores (5 nodes), and only 41 cores from each node would be used.

If these options are not feasible, I believe we could implement the balancing by trying to achieve something similar as what's done by mpirun when machinefiles are not specified. What is still not very clear is why it leaves out always the same cores (21-23 and 44-47), as Enza observed in her tests:

```
node 1: cores 0-20 24-43
node 2: cores 0-20 24-43
node 3: cores 0-20 24-43
node 4: cores 0-20 24-43
node 5: cores 0-19 24-43
```

As far as I know, with the syntax we have in the machinefiles we cannot specify which cores we want, but only how many we need from each node. In this scenario, we are not sure if implementing the balanced approach would be worth the effort. For the time being, it would be ideal to find a number multiple of 48 that can be used. Enza was going to check the feasibility with @orioljorba.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 18, 2018, 24:33

@orioljorba, do you think that we can live with having to use a multiple of 48 when we run MONARCH in a workflow that uses wrappers? this contrain means that we have to be able to decompose PROCESSORS-WRTSK=INPES*JNPES. eg if we use 4 nodes (ie 48*4=192 cores), we can only do 192-8=8*23
do you see a problem in having a more limited choice for INPES and JNPES than we currently have or is OK?
(Btw, for WRTSK I have always used either 4 or 8, can we be more flexible than that?) We need your expertise with this decision, thanks! :) 

@ lbatista, option 2 can be used already without you have to make any change? on one hand it is good to have options, the downside of it is that we add one more constraint:  PROCESSORS has to be divisible by the number of nodes e.g. PROCESSORS=14*15+8=218 would need the request of 5 nodes, but cannot be divided by 5.

**Comment by kinow** [comment]
In GitLab by @orioljorba on Jul 18, 2018, 09:31

@ editomas From past experience, MONARCH has issues if writing tasks and computing tasks are in the same node, so my recommendation is to use n*48 for computing node plus an extra node for the I/O. Now, I'm not sure if this limitation still persist, but I follow this rule.

There is no limitation to use cores for WRTSK, but the optimal number depends on the resolution and size of the model.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 18, 2018, 10:07

Hi!

First of all, sorry because I didn't provide feedback from the operational model. I told Kim to check the configuration used and they were not providing any special options, so while they thought the processes were being assigned sequentially to the nodes, in fact they were being evenly distributed (it was like 40 cores per node in the end).

Then, Kim wanted to do a test assigning the cores sequentially as it was the default behavior in MN3. This was to ensure I/O processes are separated from the computing ones, as @orioljorba describes. It was tricky but finally it could be done and the result is that the simulation takes 3 minutes more than when the load is well balanced.

About why the same cores (21-23, 44-47)  are always empty, well, the machinefile is assigning processes to nodes, but is not forcing them to be in any particular physical core. So Intel MPI tries to optimize de performance by evenly distributing the resources across the two sockets, so they have the same memory bandwith, cache, and shared resources.

So, summarizing:

* It seems that MONARCH performs better with balanced resources, and it seems mixing computing and I/O processes is not really bad for performance.
* By using machinefiles we can decide to which node the processes would go to. The distribution inside the nodes is optimized by the runtime. In any case there are options to change this behavior.
* Option 1) by @ lbatista is the most straightforward and the only one available right now, I think. For option 2) some development would be needed.
* What still is not very clear to me is why some tests @ editomas did took around 40 minutes, that is too much. The tests Kim did with load imbalance only spent 3 minutes more of wallclock time.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 18, 2018, 11:15

@mcastril thanks for the update about what you have discovered. 

About your last point, Kim has spent 3 minutes relative to which time? The tests I had done were high resolution regional simulations with relatively few cores, so, in standard conditions, they take about 17 minutes. I suspect that the test of Kim was much shorter and so the relative increase of time. 

To be precise, to the 2 options you mentioned we should add option 3) also by @ lbatista "If these options are not feasible, I believe we could implement the balancing by trying to achieve something similar as what's done by mpirun when machinefiles are not specified.", for which I believe more development is needed than option 2).

We'll talk internally and let you know if we can live with option 1.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 18, 2018, 11:49

Hi @ editomas , and sorry for the confusion, in fact I thought 2) was 3). For 2) I don't know if some development is required.

About Kim's run, he told me that by using `#SBATCH --distribution=plane=48` to force using 48 cores per node (the one would use only 8 as there are no more cores to put), total time is 34m39.224s , which is 3 minutes more than usual, according to him. So the difference in relative terms is much lower than yours.

On the other hand, in his case (48x48x48x48x48x8) the computing processes are still equally distributed, which was not your case (36x48x48x24). In his case the small difference could come from the fact that he is filling the nodes, and therefore leaving less resources to each process. In your runs (those that were very slow) there was a lot of imbalance between computing processes. Of course there are many obscure things to me, but this is the logic I can do from all the evidence. I think there's something inside MONARCH that reacts badly to having that imbalance.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 19, 2018, 18:13

@mcastril I had a chat with @orioljorba, for the time being we will keep this reduced flexibility in the number of processors that we can use for MONARCH (option 1) without asking for more development. 

Once/if we understand better this behaviour of MONARCH's efficiency, if you are willing, we might ask you to enhance autosubmit to write the machine files balancing the nodes better than it does currently. How does it sound?

Regarding your last comment, it is interesting what you have point out, that what might matter more is the imbalance between computing processes only, and that this could explain why Kim's test was only 3 minutes slower. We could test this point (i.e. PROCESSORS= n*48+8 (where n is any integer) and compare it to PROCESSORS= n*48).

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 20, 2018, 18:18

I think it is a good plan @ editomas . We should validate my logic and see if that is what the model need, that would save some effort now that Larissa has to take care of many things (documentation, closing issues, etc).

Thanks!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jan 2, 2020, 11:22

closed
